### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by panianchuk**
https://panianchuk.github.io/react_todo-app-add-and-delete/#/

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,223 +1,162 @@
...
+  const [error, setError] = useState('');
+  const [isHiddenErrorNote, setIsHiddenErrorNote] = useState(true);
+  const [isAdding, setIsAdding] = useState(false);
   const newTodoField = useRef<HTMLInputElement>(null);
+  const visibleTodos = filterTotos(todos, typeFilter);
+  const activeTodosCount = filterTotos(todos, FilterTypes.Active).length;
+  const completedTodosCount = filterTotos(todos, FilterTypes.Completed).length;
+  const allCompleted = todos.length === completedTodosCount;
+
+  function hiddenErrorNote() {
```
Comment (Volodymyr-Melnyk-Fron10der) [inline src/App.tsx:26]:
```suggestion
  function showErrorNote() {
```

File: src/App.tsx [hunk 2]
```diff
@@ -1,223 +1,162 @@
...
 
 export const App: React.FC = () => {
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
-  const user = useContext(AuthContext);
+  const user = useAuthContext();
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [title, setTitle] = useState('');
+  const [typeFilter, setTypeFilter] = useState(FilterTypes.All);
+  const [error, setError] = useState('');
+  const [isHiddenErrorNote, setIsHiddenErrorNote] = useState(true);
```
Comment (Volodymyr-Melnyk-Fron10der) [inline src/App.tsx:18]: It is difficult logic. you have hidden error by default, so it's better to name variable isErrorNoteShown and set it as false. And only when you need to show the error you do setErrorNoteShown(true) - so now it really do what it suppose to do

File: src/api/filter.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+export function filterTotos(todosArr: Todo[], type: string) {
+  switch (type) {
+    case FilterTypes.Completed:
+      return todosArr.filter((todo) => todo.completed);
+    case FilterTypes.Active:
+      return todosArr.filter((todo) => !todo.completed);
+    default:
+      return todosArr;
+  }
+}
```
Comment (Volodymyr-Melnyk-Fron10der) [inline src/api/filter.ts:12]: Move it into `helpers `folder

File: src/App.tsx [hunk 4]
```diff
@@ -1,223 +1,162 @@
...
+  const [isAdding, setIsAdding] = useState(false);
```
Comment (Volodymyr-Melnyk-Fron10der) [inline src/App.tsx:19]: isLoading / isCreating would sound such better

File: src/App.tsx [hunk 5]
```diff
@@ -1,223 +1,162 @@
...
   const newTodoField = useRef<HTMLInputElement>(null);
+  const visibleTodos = filterTotos(todos, typeFilter);
+  const activeTodosCount = filterTotos(todos, FilterTypes.Active).length;
+  const completedTodosCount = filterTotos(todos, FilterTypes.Completed).length;
+  const allCompleted = todos.length === completedTodosCount;
```
Comment (Volodymyr-Melnyk-Fron10der) [inline src/App.tsx:24]: use useMemo for better performance

File: src/App.tsx [hunk 6]
```diff
@@ -1,223 +1,162 @@
...
+
+    if (user) {
+      getTodos(user.id)
+        .then((res) => setTodos(res))
+        .catch(() => {
+          setError(ErrorTypes.Loading);
+          hiddenErrorNote();
+        })
+        .finally(() => setIsAdding(false));
+    }
```
Comment (Volodymyr-Melnyk-Fron10der) [inline src/App.tsx:48]: If there is no user your `isAdding ` will be always true, fix it

File: src/App.tsx [hunk 7]
```diff
@@ -1,223 +1,162 @@
...
+      .catch(() => {
+        setError(ErrorTypes.Remove);
+        setTimeout(() => {
+          setIsHiddenErrorNote(true);
+        }, 3000);
+      });
+  };
+
+  const deleteCompletedTodosHandler = () => {
+    todos.map((todo) => (todo.completed ? deleteTodoHandler(todo.id) : todo));
```
Comment (Volodymyr-Melnyk-Fron10der) [inline src/App.tsx:98]: Why map? Just use forEach

**Review by Volodymyr-Melnyk-Fron10der — CHANGES_REQUESTED** [review]
Hi, I left comments +

1. move your useAuthContext into `src -> hooks`. Why it is in components folder now?)
2. Choose one method of organazing components: all are just the files | all are folders with files inside
image

File: src/components/useAuthContext.ts [hunk 8]
```diff
@@ -0,0 +1,6 @@
+import { useContext } from 'react';
+import { AuthContext } from '../hooks/AuthContext';
+
+export const useAuthContext = () => {
+  return useContext(AuthContext);
+};
```
Comment (DarkMistyRoom) [inline src/components/useAuthContext.ts:6]: better import it from the context folder, to avoid confusion. it isn't a component

**Review by DarkMistyRoom — APPROVED** [review]

**Pull request closed without merging**